### ISSUE TITLE
Standalone library?

### ISSUE THREAD

**Issue by ainsley** [eclipse/tm4e#40]
Not strictly an issue but writing to ask if this library is available in a standalone "core" version without any dependencies on Eclipse. I really would like to see this integrated into a Swing or FX editor pane. I saw that there was an issue raised on the RSyntaxTextArea github issues page regarding this, but it appears that the core project still contains Eclipse PDT dependencies.

Chris

**Comment by angelozerr** [comment]
@ainsley at first textmate.java (called now tm4e) is migrating to Eclipse.org https://projects.eclipse.org/projects/technology.tm4e

Even if the Core is linked to Eclipse API, you can consume the textmate API https://github.com/angelozerr/textmate.java/wiki/Core without Eclipse (just get the JAR).

I'm waiting for people like RSyntaxTextArea guys are interested to improve the documentation.

**Comment by ainsley** [comment]
Excellent. Is there a place for jar releases at present? And are there any external dependencies? I see there may be a JRuby dependency (which is quite large), is that a hard dependency?

**Comment by angelozerr** [comment]
> Is there a place for jar releases at present?

Yes at http://oss.opensagres.fr/textmate/0.1.0-SNAPSHOT/plugins/

> And are there any external dependencies?

Yes it is linked to Joni because textmate needs onigurama. You can find the two JARS at https://github.com/angelozerr/textmate.java/tree/master/thirdparties/joni-osgi (jcodings-1.0.13.jar and joni-2.1.11.jar)

But as I said you, tm4e will be move to Eclipse.org. I need to discuss with @mickaelistria how we could provide an easy mean to consume tm4e outside Eclipse project (like RSyntaxTextArea ) when people will be interested. Perhaps by register tm4e in the maven repository central?

**Comment by ainsley** [comment]
Hi,

I have tried build the "core" library outside Eclipse (also including the two jars you mentioned). Trying to do it this way produces 379 build errors. There are about 10 errors in TMCorePlugin, which I would expect from an Eclipse plugin project, but there are many many more based on unresolved library dependencies beyond the two jars.

Seems that there are more dependencies:

org.w3c.css.sac..*;
om.google.gson.stream.*; (I think for JSon Reader, which I would imagine is optional to the core).

Possibly more.

It's a good idea to have a core library then a larger Eclipse Plugin library, but as it stands, it's not really portable outside of Eclipse without manually making a number of modifications.

As this is a lot of work, I would not dare to presume this is anywhere near your list of priorities, but I think it worth pointing out that it is possible to modularise this some more.

**Comment by ainsley** [comment]
FYI, I tried to resolve the "org.w3c.css.sac..;" dependencies by including in the "sac-1.3.jar" which appear to be stubs for CSS data structures. This reduces the number of errors down to 54. Those errors are in two categories:

1) Eclipse Plugin Bindings. I think these are safe to ignore if referencing from say Swing or Java FX.
2) JSONPListParser class: I also think this is safe to ignore, if we will not be using JSON based PLists.

As such, I'll probably delete the classes with errors manually and try to see what I can do to unofficially integrate with JSyntaxTextArea, if I can figure out both APIs.

**Comment by angelozerr** [comment]
@ainsley I understand that today it's not an easy task. I suggest you that you work with JAR. 

Please let us time to discuss about that with @mickaelistria

**Comment by ainsley** [comment]
No problem. I'll see what I can do on the RSyntaxText area integration. If when you have time, you could update the API documentation in any way you imagine could be useful it would be a huge help. If you don't have time, no problem, I'll try to figure it out myself.

**Comment by angelozerr** [comment]
I suggest you 

1) you implement your own TM model https://github.com/angelozerr/textmate.java/blob/master/org.eclipse.tm4e.ui/src/main/java/org/eclipse/tm4e/ui/internal/model/TMModel.java (update the TM model according changes of your editor)

2) you implement your own class like https://github.com/angelozerr/textmate.java/blob/master/org.eclipse.tm4e.ui/src/main/java/org/eclipse/tm4e/ui/text/TMPresentationReconciler.java#L348 which intercept changes of TM Model to get TM tokens and colorize it as you wish (I suggest you to use TextMate theme (CSS is here for that, but I would like to support TextMate theme too)

Good luck! Any feedback and PR are welcome and I know that AbstractTMModel should be a lot improved.

**Comment by ainsley** [comment]
Thanks for the info. Excellent project.

**Comment by angelozerr** [comment]
Thanks @ainsley but please note that this project must be improved again with AbstractTMModel to synch model with editor with clean mean.

**Comment by ainsley** [comment]
It looks like RSyntaxTextArea is using a very different way of performing highlighting. It essentially updates lines based on the region of text changed, but I don't believe it buffers on a timeout, so lots of events are fired. Worse still, it only knows to update the line after an affected line if the end of line token has changed from the previous end of line token (I may be wrong here though). So essentially, you have a shallow token stack of one entry.

Long story short. I'll have to put this on the backburner for a while.

**Comment by angelozerr** [comment]
> It looks like RSyntaxTextArea is using a very different way of performing highlighting,

Is it different from Eclipse?

**Comment by ainsley** [comment]
Please bear in mind, I'm not an expert on RSyntaxTextArea so I'm quite open to getting this analysis wrong, but with that said, this is what I now believe.

RSyntaxTextArea uses line based dirty region syntax highlighters, but essentially and the highlighter has priority for watching document deltas. Not only that, but it appears that the document delta region is based on the dirty region and floods over to subsequent lines by looking at the end token of a delta line for different. Essentially maintaining its own state stack that can only hold one item on the stack. That is, if a delta is applied to a document, such that it affects only one line, but after recalculating the tokens for that line, the default token at the end of the line changes (for example typing a "/*" in a c style language, then each subsequent line needs to be re-tokenized/colorized until there are no changes in the end token on subsequent lines. Such an approach is efficient within its constraints but I'm not sure how to plug in "textmate.java" into this world.

I thought about maintaining a virtual document with the "textmate.java" tokens, by using my own document listener, then using those tokenized text blocks in the main per-line colorization block. But, the colorization block has priority for document deltas, therefore I cannot colorize without some kind of deferred callback - such a callback requiring non trivial adjustments to the framework.

Now, I could be completely wrong about some or all of this. I'm a non-expert 3rd party looking at two components I think the correct approach is to let the author of RSyntaxTextArea evaluate how best to integrate it into the library themselves, regrettably. It's a great library, and I use it already using a JFlex definition, but it also suffers from a shallow stack. Textmate grammars are so much more flexible. 

The search continues for a swing or fx based textmate-grammar capable syntax highlighter.

**Comment by mickaelistria** [comment]
@ainsley This project is currently mainly targetting Eclipse and OSGi, so the jars are offered in the p2 format and dependencies are defined in the OSGi way rather than Maven for the moment. That's definitely something worth improving.
As a temporary workaround, you can find in https://github.com/angelozerr/textmate.java/blob/master/org.eclipse.tm4e.core/META-INF/MANIFEST.MF the required-bundle and import-packages for the core (Eclipse-independent) parts of TM4E. They represent the (top-level) dependencies of the projet. Resolving those and their transitive deps as Maven artifacts should provide you a working classpath for TM4E.

**Comment by ainsley** [comment]
Might I suggest creating an additional, non-Eclipse project, with the core library MINUS the Eclipse PDE classes? The core library could still be Maven and OSGi compliant, but would not require any sort of Eclipse dependency.

There is no great urgency for this, as I appear to hit a brick wall integrating the library into RSyntaxTextArea as described in an earlier post. It just seems a more modular approach.

**Comment by mickaelistria** [comment]
[The project is moving under the eclipse.org umbrella.](https://projects.eclipse.org/projects/technology.tm4e) Let's wait for migration to be complete before refactoring it. I think it's better to keep it a single project to not scatter workforce, but have this project releasing the core part as Maven artifact. Eclipse Platform and JDT managed to do it properly recently ( https://objectteams.wordpress.com/2017/01/09/eclipse-neon-2-is-on-maven-central/ ), so we can probably make it happen for TM4E too.
We can discuss this in a separate issue, but since Swing and JavaFX are supposed to be part of the Java distribution, it might make sense to have your integrations of TM4E with those graphical APIs part of the project directly.

**Comment by mickaelistria** [comment]
At the moment, the `org.eclipse.tm4e.core.internal.grammars` requires Eclipse APIs as it deal with the extension registry and content-type->grammar resolution. If we want to keep concerns separated (Textmate processing vs Eclipse IDE rendering), this package should probably move to the `org,eclipse.tm4e.ui` bundle which is about Eclipse IDE integration. Then `org.eclipse.tm4e.core` could be used in any Java app.

**Comment by angelozerr** [comment]
An other idea is to have 3 plugins like:

 * org.eclipse.tm4e.commons which is not linked to Eclipse API 
 * org.eclipse.tm4e.core which hosts org.eclipse.tm4e.core.internal.grammars
 * org.eclipse.tm4e.ui like today

Havings org.eclipse.tm4e.core.internal.grammars packages inside the UI plugins disturbs me a little because it is not linked to UI.

@mickaelistria what do you think about that?

**Comment by mickaelistria** [comment]
Is there any use-case of `org.eclipse.tm4e.core.internal.grammars` which isn't targetting UI ? If it's used only for Eclipse UI, it's IMHO better to have it directly in the UI plugin.

**Comment by angelozerr** [comment]
The main goal of `org.eclipse.tm4e.core.grammars` is to link `org.eclipse.core.runtime.content.IContentType` which belongs to Eclipse Core plugin with `org.eclipse.tm4e.core.grammar.IGrammar` and that's all.

There is no UI like SWT widgets or other things for `org.eclipse.tm4e.core.grammars`.

**Comment by mickaelistria** [comment]
Ok, so I suggest the 3 bundles like you did earlier but with different names:
* org.eclipse.tm4e.core, for the grammar processing (doesn't depend on org.eclipse package)
* org.eclipse.tm4e.registry, for the registry and content-type->grammar association (doesn't depend on Eclipse UI)
* org.eclipse.tm4e.ui, for the integration in Eclipse IDE UI

WDYT?

**Comment by angelozerr** [comment]
> WDYT?

Good idea!

**Comment by mickaelistria** [comment]
The `org.eclipse.tm4e.core` bundle has been made a plain Java library without any dependency to Eclipse APIs nor OSGi.
A next step could be to publish it on some Nexus repo (such as repo.eclipse.org). If this happens to be useful, please open a new ticket.
